src/com/amazon/ion/impl/IonCursorBinary.java

```diff
@@ -0,0 +1,721 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.amazon.ion.impl;
+
+import com.amazon.ion.BufferConfiguration;
+import com.amazon.ion.IonBufferConfiguration;
+import com.amazon.ion.IonException;
+import com.amazon.ion.IonCursor;
+import com.amazon.ion.IonType;
+import com.amazon.ion.IvmNotificationConsumer;
+
+import java.nio.ByteBuffer;
+
+/**
+ * An IonCursor over binary Ion data, capable of buffering or skipping Ion values at any depth. Records byte
+ * indices of the value currently buffered, enabling parsers direct access to the bytes in the value's representation.
+ */
+class IonCursorBinary implements IonCursor {
+
+    private static final int LOWER_SEVEN_BITS_BITMASK = 0x7F;
+    private static final int VALUE_BITS_PER_VARUINT_BYTE = 7;
+    private static final int IVM_START_BYTE = 0xE0;
+    private static final int IVM_FINAL_BYTE = 0xEA;
+    private static final int SINGLE_BYTE_MASK = 0xFF;
+    private static final int LIST_TYPE_ORDINAL = IonType.LIST.ordinal();
+    private static final IvmNotificationConsumer NO_OP_IVM_NOTIFICATION_CONSUMER = (x, y) -> {};
+
+    // Initial capacity of the stack used to hold ContainerInfo. Each additional level of nesting in the data requires
+    // a new ContainerInfo. Depths greater than 8 will be rare.
```

Comment:
```suggestion
    // a new ContainerInfo. Depths greater than 8 are assumed to be rare.
```

Reply:
Addressed in https://github.com/amazon-ion/ion-java/pull/538/files